# Analyzing Worldwide and Regional and CPI Score Changes from 2012 to 2021 

## Introduction

Corruption, at least to me, is one those things we want to eradicate yet it seems that no progress has been made. This assessment of "no progress" is likely based on seeing frequent news about politicians caught being corrupt or in some instances, politicians who were not caught but free to do as they please in spite of the evidence. I'm interested in seeing whether the data on corruption supports the view that unfortunately, no progress has been made on eradicating or more feasibly, reducing corruption.

Before that, I need to clarify the definition of corruption and how it impacts the questions to answer. The simplest definition comes from Transparency International, an institution dedicated to ending worldwide corruption, which defines corruption as ["the abuse of entrusted power for private gain"](https://www.transparency.org/en/what-is-corruption). But that simple definition might hide the fact that corruption encompassess multiple activities and more importantly, difficult to precisely compare i.e. generally, we might say person A is more corrupt than person B but can we precisely point the exact difference in corruption between the two, for example, in terms of numbers? 

Incidentally, to make it easier to compare corruption level and unify the multiple corruption measurements available, Transparecy International created the [Corruption Perception Index](https://www.transparency.org/en/cpi/) which scores a country's corruption based on ["at least 3 data sources drawn from 13 different corruption surveys and assessments"](https://www.transparency.org/en/news/how-cpi-scores-are-calculated). The CPI Index scores a country from 0 to 100 and the higher the value, the least corrupt the country is perceived to be. It's important to note what is meant by "Perception" in the index's name. The CPI only [measures how corrupt each country’s public sector is perceived to be, according to experts and businesspeople](https://www.transparency.org/en/news/how-cpi-scores-are-calculated); it doesn't measure citizens' direct perceptions or experience of corruption which is measured by a different measurement called the [Global Corruption Barometer](https://www.transparency.org/en/gcb) which is also created by Transparency International.

By choosing the CPI Index as representative of corruption, the questions guiding this project will be as follows:

1. How has the worldwide average CPI score changed from 2012 to 2021?
2. Is the worldwide average CPI score in 2021 higher than in 2012?
3. How has the regional average CPI score changed from 2012 to 2021?
4. How many regions average CPI score are higher in 2021 than in 2012?

### Data Sets

The data sets used in this project can be downloaded from the [CPI](https://www.transparency.org/cpi) pages of the Transparency International website. Note there are 10 different pages with each corresponding to a year's data set. Other than that, some of the pages don't have direct links to the data sets and the data sets are stored in a zip file together with other data sets which are not relevant to this project. Also important is that a file might contain multiple data sets or sheets/tables; most of the time, this is because the file contains both the worldwide and regional CPI scores in a single file. In this project, I'll only focus on the worldwide CPI data set thus the regional scores will be removed.

The links for each year's CPI page followed by brief instructions on getting the relevant data set:

**2012**

1. Visit https://www.transparency.org/cpi/2012.
2. Under the Full 2012 Materials section, click the download button next to "CPI 2012 Data Set". [Direct download link](https://images.transparencycdn.org/images/CPI2012_Results.xls). 
3. There's only one file — `CPI2012_Results.xls`. I'll only use the worldwide `CPI 2012` sheet and not the regional sheets.

**2013**

1. Visit https://www.transparency.org/en/cpi/2013.
2. Under the Full 2013 Materials section, click the download button next to "CPI2013 Results". [Direct download link](https://images.transparencycdn.org/images/CPI2013_Results_2022-01-20-183035_stnh.xlsx).
3. There's only one file — `CPI2013_Results_2022-01-20-183035_stnh.xlsx` — and its only sheet `CPI 2013` fits the requirement of the project.

**2014**

1. Visit https://www.transparency.org/en/cpi/2014.
2. Under the Full 2014 Materials section, download the "CPI 2014 Data Set". [Direct download link](https://images.transparencycdn.org/images/CPI2014_Results.xlsx).
3. There's only one file — `CPI2014_Results.xlsx`. I'll only use the worldwide `CPI 2014` sheet and not the regional sheets.

**2015**

1. Visit https://www.transparency.org/en/cpi/2015.
2. Under the Full 2015 Materials section, click the download button next to "CPI 2015 Full Data Set 2022 01 18 145020 enyn". [Direct download link](https://images.transparencycdn.org/images/CPI_2015_FullDataSet_2022-01-18-145020_enyn_2022-01-20-180010_mabu.xlsx).
3. There's only one file – `CPI_2015_FullDataSet_2022-01-18-145020_enyn_2022-01-20-180010_mabu.xlsx`. I'll only use the worldwide `CPI 2015` sheet and not the regional sheets.

**2016**

1. Visit https://www.transparency.org/en/cpi/2016.
2. Under the Full 2016 Materials section, click the download button next to "CPI 2016 Data set". [Direct download link](https://images.transparencycdn.org/images/CPI2016_Results.xlsx).
3. There's only one file – `CPI2016_Results.xlsx`. I'll only use the worldwide `CPI2016_FINAL_16Jan` sheet and not the regional sheets nor the `CPI 2015-2016` sheet.

**2017**

1. Visit https://www.transparency.org/en/cpi/2017.
2. Under the Full 2017 Materials section, click the download button next to "CPI2017 Full Data Set". [Direct download link](https://images.transparencycdn.org/images/CPI2017_Full_DataSet-1801.xlsx).
3. There's only one file – `CPI2017_Full_DataSet-1801.xlsx`. I'll only use the worldwide `CPI 2017` sheet and not the regional sheets.

**2018**

1. Visit https://www.transparency.org/en/cpi/2018.
2. Under the Full 2018 Materials section, click the download button next to "CPI 2018 Full Results". [Direct download link](https://images.transparencycdn.org/images/CPI2018_Full-Results_1801.xlsx).
3. There's only one file – `CPI2018_Full-Results_1801.xlsx`. I'll only use the worldwide `CPI2018` sheet and not the regional sheets.

**2019**

1. Visit https://www.transparency.org/en/cpi/2019.
2. Under the Full 2019 Materials section, click the download button next to "CPI Full Data Set". [Direct download link](https://images.transparencycdn.org/images/CPI2019-1.xlsx).
3. There's only one file – `CPI2019-1.xlsx`. I'll only use the worldwide `CPI2019` sheet and not the regional sheets.

**2020**

1. Visit https://www.transparency.org/en/cpi/2020.
2. Under the Full 2020 Materials section, click the download button next to "CPI Full Data Set (.zip)". [Direct download link](https://images.transparencycdn.org/images/CPI_FULL_DATA_2021-01-27-162209.zip).
3. Unzip the folder.
4. There are two files – `CPI2020_GlobalTablesTS_210125.xlsx` and `CPI2020_SignificantChanges_210125.xlsx`. I'll be using the former. In that file, I'll only use the worldwide `CPI2020` sheet and not the regional sheets.

**2021**
1. Visit https://www.transparency.org/en/cpi/2021.
2. Under the Full 2021 Materials section, click the download button next to "CPI Full Data Set". [Direct download link](https://images.transparencycdn.org/images/CPI-2021-Full-Data-Set.zip).
3. Unzip the folder.
4. There are two files – `CPI2021_GlobalResults&Trends.xlsx` and `CPI2021_RegionalTables.xlsx`. I'll be using the former. In that file, I'll only use the worldwide `CPI 2021` sheet and not the regional sheets.


*Caveat: The above instructions are correct as of the 10th of October 2022.*

## Data Cleaning: Merging All Data Sets

The data sets structure is not consistent from one set to another. Aside from that, analyzing 10 separate data sets instead of just one can be tedious and error-prone.

Next, I'll go through each data set and merge two data sets at a time. Aside from that, to reduce the data cleaning work in the final data set, I'll also review some of the data set properties such as number of rows and columns along the way, and then, clean the data as needed.

### Preparing Year 2021

Let's start by loading the year 2021 file.

In [1]:
import pandas as pd
# Read from the excel file; ignore the first two rows because they are not headers
CPI_2021 = pd.read_excel("raw_datasets/CPI2021_GlobalResults&Trends.xlsx", sheet_name="CPI 2021", header=2)

# Review the data set
print(CPI_2021.shape)
CPI_2021

(180, 22)


Country / Territory ISO3 Region  CPI score 2021  Rank  Standard error  \
0               Denmark  DNK  WE/EU              88     1        1.850507   
1               Finland  FIN  WE/EU              88     1        1.248280   
2           New Zealand  NZL     AP              88     1        1.427644   
3                Norway  NOR  WE/EU              85     4        1.072085   
4             Singapore  SGP     AP              85     4        1.159654   
..                  ...  ...    ...             ...   ...             ...   
175               Yemen  YEM   MENA              16   174        2.292554   
176           Venezuela  VEN    AME              14   177        0.991130   
177             Somalia  SOM    SSA              13   178        1.891675   
178               Syria  SYR   MENA              13   178        1.578473   
179         South Sudan  SSD    SSA              11   180        0.409076   

     Number of sources   Lower CI  Upper CI  African Development Bank CPIA  \
0                    8  84.955920  91.04408                            NaN   
1                    8  85.946580  90.05342                            NaN   
2                    8  85.651530  90.34847                            NaN   
3                    7  83.236420  86.76358                            NaN   
4                    9  83.092370  86.90763                            NaN   
..                 ...        ...       ...                            ...   
175                  7  12.228750  19.77125                            NaN   
176                  8  12.369590  15.63041                            NaN   
177                  6   9.888195  16.11180                           11.0   
178                  5  10.403410  15.59659                            NaN   
179                  5  10.327070  11.67293                           11.0   

     ...  Economist Intelligence Unit Country Ratings  \
0    ...                                         90.0   
1    ...                                         90.0   
2    ...                                         90.0   
3    ...                                         90.0   
4    ...                                         90.0   
..   ...                                          ...   
175  ...                                         20.0   
176  ...                                         20.0   
177  ...                                          NaN   
178  ...                                         20.0   
179  ...                                          NaN   

     Freedom House Nations in Transit  Global Insights Country Risk Ratings  \
0                                 NaN                                    83   
1                                 NaN                                    83   
2                                 NaN                                    83   
3                                 NaN                                    83   
4                                 NaN                                    83   
..                                ...                                   ...   
175                               NaN                                    10   
176                               NaN                                    10   
177                               NaN                                    22   
178                               NaN                                    10   
179                               NaN                                    10   

     IMD World Competitiveness Yearbook  PERC Asia Risk Guide  \
0                                  93.0                   NaN   
1                                  94.0                   NaN   
2                                  91.0                   NaN   
3                                  81.0                   NaN   
4                                  87.0                  91.0   
..                                  ...                   ...   
175                                 NaN              

Note that for the more recent years, Transparency International standardized the number of countries for their worldwide ranking to the top 180 countries in terms of CPI score. Thus, the 180 number of rows is as expected.

Looking at the columns, I can see that not all of them are relevant to this project. I'll only keep the following columns:

1. Country/Territory
2. ISO3
3. Region
4. CPI score 2021
5. Rank

In [2]:
# Pick the first 5 columns
CPI_2021 = CPI_2021.iloc[:, :5]

# Review modified data frame
CPI_2021

Country / Territory ISO3 Region  CPI score 2021  Rank
0               Denmark  DNK  WE/EU              88     1
1               Finland  FIN  WE/EU              88     1
2           New Zealand  NZL     AP              88     1
3                Norway  NOR  WE/EU              85     4
4             Singapore  SGP     AP              85     4
..                  ...  ...    ...             ...   ...
175               Yemen  YEM   MENA              16   174
176           Venezuela  VEN    AME              14   177
177             Somalia  SOM    SSA              13   178
178               Syria  SYR   MENA              13   178
179         South Sudan  SSD    SSA              11   180

[180 rows x 5 columns]

Before moving to the next data set, I'm going to rename some of the columns for easier references later.

In [3]:
# Rename some of the columns
CPI_2021.rename(columns={ "Country / Territory": "Country", "Rank": "Rank 2021", "CPI score 2021": "CPI Score 2021"}, 
                inplace=True)

CPI_2021.columns

Index(['Country', 'ISO3', 'Region', 'CPI Score 2021', 'Rank 2021'], dtype='object')

### Preparing Year 2020 and Merging with Year 2021

#### Load the year 2020 data set and select relevant columns

Unfortunately, for unknown reasons, the year 2020 file — `raw_datasets/CPI2020_GlobalTablesTS_210125.xlsx` — can't be loaded by the same function I used for year 2021. It could be a [malformed](https://github.com/pandas-dev/pandas/issues/39250#issuecomment-762380350) `.xlsx` file but I can't say for sure.

The simplest fix is to save the file as an `.xls` file using a spreadsheet editor. I'm using the Numbers software on my Mac and here's a [general guide](https://support.apple.com/en-my/guide/numbers/tan3b922d4ad/mac) on converting a file into a different format. In Excel, this [answer](https://answers.microsoft.com/en-us/msoffice/forum/all/converting-xlsx-workbook-to-xls-format-in-excel/ece28ab5-1386-4421-b424-dba673a75708) on the Microsoft forum could be useful. Another option is to download the file I've processed on my computer: here's a [link](https://github.com/wanzulfikri/cpi_analysis/blob/main/processed_datasets/CPI2020_GlobalTablesTS_210125%202.xls) to the processed file after converting it with Numbers.

In the end, the new file will be named `CPI2020_GlobalTablesTS_210125.xls`.

In [4]:
CPI_2020 = pd.read_excel("processed_datasets/CPI2020_GlobalTablesTS_210125.xls", sheet_name="CPI2020", header=2)

# Review the data set
print(CPI_2020.shape)
CPI_2020

(180, 22)


Country ISO3 Region  CPI score 2020  Rank  Standard error  \
0        Denmark  DNK  WE/EU              88     1        1.775809   
1    New Zealand  NZL     AP              88     1        1.479342   
2        Finland  FIN  WE/EU              85     3        1.748594   
3      Singapore  SGP     AP              85     3        1.203239   
4         Sweden  SWE  WE/EU              85     3        1.303953   
..           ...  ...    ...             ...   ...             ...   
175    Venezuela  VEN    AME              15   176        0.914430   
176        Yemen  YEM   MENA              15   176        1.333599   
177        Syria  SYR   MENA              14   178        1.576392   
178      Somalia  SOM    SSA              12   179        2.292273   
179  South Sudan  SSD    SSA              12   179        1.108644   

     Number of sources   Lower CI  Upper CI  African Development Bank CPIA  \
0                    8  85.087670  90.91233                            NaN   
1                    8  85.573880  90.42612                            NaN   
2                    8  82.132310  87.86769                            NaN   
3                    9  83.026690  86.97331                            NaN   
4                    8  82.861520  87.13848                            NaN   
..                 ...        ...       ...                            ...   
175                  8  13.500340  16.49966                            NaN   
176                  7  12.812900  17.18710                            NaN   
177                  5  11.414720  16.58528                            NaN   
178                  6   8.240673  15.75933                            4.0   
179                  5  10.181820  13.81818                           11.0   

     ...  Economist Intelligence Unit Country Ratings  \
0    ...                                         90.0   
1    ...                                         90.0   
2    ...                                         72.0   
3    ...                                         90.0   
4    ...                                         90.0   
..   ...                                          ...   
175  ...                                         20.0   
176  ...                                         20.0   
177  ...                                         20.0   
178  ...                                          NaN   
179  ...                                          NaN   

     Freedom House Nations in Transit  Global Insight Country Risk Ratings  \
0                                 NaN                                   83   
1                                 NaN                                   83   
2                                 NaN                                   83   
3                                 NaN                                   83   
4                                 NaN                                   83   
..                                ...                                  ...   
175                               NaN                                   10   
176                               NaN                                   10   
177                               NaN                                   10   
178                               NaN                                   22   
179                               NaN                                   10   

     IMD World Competitiveness Yearbook  PERC Asia Risk Guide  \
0                                  94.0                   NaN   
1                                  92.0                   NaN   
2                                  91.0                   NaN   
3                                  91.0                  90.0   
4                                  84.0                   NaN   
..                                  ...                   ...   
175                                20.0                   NaN   
176                                 NaN                   NaN   
177                     

Similar to before, the number of rows match the 180 country limit, and the headers and values look fine at least on a cursory look.

I'll also keep the following columns and discard the rest:

1. Country
2. ISO3
3. Region
4. CPI score 2020
5. Rank

Plus, renaming as usual.

In [5]:
# Pick the first 5 columns
CPI_2020 = CPI_2020.iloc[:, :5]

# Rename some of the columns
CPI_2020.rename(columns={ "Rank": "Rank 2020", "CPI score 2020": "CPI Score 2020"}, 
                inplace=True)

# Review modified data frame
CPI_2020

Country ISO3 Region  CPI Score 2020  Rank 2020
0        Denmark  DNK  WE/EU              88          1
1    New Zealand  NZL     AP              88          1
2        Finland  FIN  WE/EU              85          3
3      Singapore  SGP     AP              85          3
4         Sweden  SWE  WE/EU              85          3
..           ...  ...    ...             ...        ...
175    Venezuela  VEN    AME              15        176
176        Yemen  YEM   MENA              15        176
177        Syria  SYR   MENA              14        178
178      Somalia  SOM    SSA              12        179
179  South Sudan  SSD    SSA              12        179

[180 rows x 5 columns]

#### Merge year 2021 dataset with the year 2020 dataset

To merge the two, I'll match the rows of each dataset on the `ISO3` column; this is because each country's ISO3 code is generally consistent in all of the data sets. Pay special attention to my use of "generally" because there are some pecularities with the ISO3 column in the older data sets; nevertheless, I'll handle those when they come.

An important thing to note is there might be instances in which one country is available in one data set but missing in another. To preserve those countries, I'll include them in the final data set but they'll be missing CPI score for at least one year. This is inevitable considering that even if a country has a CPI score (which can be found in supplementary data sets), it'll not be included in the global data set because of the 180 country limit per year.

For the purpose of this project, I'll only include data from the global data set only. Word of caution: this limits the representativity of the analysis results to only those countries that happen to be in the global list. Noting that limitation is essential because I'll also calculate the regional CPI score and that score might be inflated or deflated depending on which countries in the region are not included in the global data set. Since countries with very low CPI score will be excluded because their rank is higher than 180, the analysis results will likely overestimate worlwide and regional CPI score i.e. there is a bias towards countries which score high.

Now that the crucial limitation is out in the open, I'll proceed with the merging.

In [6]:
# CPI will store the CPI data for all years
CPI = CPI_2021.merge(CPI_2020, how="outer", on=["ISO3"], suffixes=["", "_2020"])

CPI

Country ISO3 Region  CPI Score 2021  Rank 2021       Country_2020  \
0        Denmark  DNK  WE/EU            88.0        1.0            Denmark   
1        Finland  FIN  WE/EU            88.0        1.0            Finland   
2    New Zealand  NZL     AP            88.0        1.0        New Zealand   
3         Norway  NOR  WE/EU            85.0        4.0             Norway   
4      Singapore  SGP     AP            85.0        4.0          Singapore   
..           ...  ...    ...             ...        ...                ...   
176    Venezuela  VEN    AME            14.0      177.0          Venezuela   
177      Somalia  SOM    SSA            13.0      178.0            Somalia   
178        Syria  SYR   MENA            13.0      178.0              Syria   
179  South Sudan  SSD    SSA            11.0      180.0        South Sudan   
180          NaN  BRN    NaN             NaN        NaN  Brunei Darussalam   

    Region_2020  CPI Score 2020  Rank 2020  
0         WE/EU            88.0        1.0  
1         WE/EU            85.0        3.0  
2            AP            88.0        1.0  
3         WE/EU            84.0        7.0  
4            AP            85.0        3.0  
..          ...             ...        ...  
176         AME            15.0      176.0  
177         SSA            12.0      179.0  
178        MENA            14.0      178.0  
179         SSA            12.0      179.0  
180          AP            60.0       35.0  

[181 rows x 9 columns]

Lastly, I'll remove the duplicate Country and Region columns.

In [7]:
# Remove duplicate Country and Region columns
CPI.drop(columns=["Country_2020", "Region_2020"],inplace=True)

CPI.head()

Country ISO3 Region  CPI Score 2021  Rank 2021  CPI Score 2020  \
0      Denmark  DNK  WE/EU            88.0        1.0            88.0   
1      Finland  FIN  WE/EU            88.0        1.0            85.0   
2  New Zealand  NZL     AP            88.0        1.0            88.0   
3       Norway  NOR  WE/EU            85.0        4.0            84.0   
4    Singapore  SGP     AP            85.0        4.0            85.0   

   Rank 2020  
0        1.0  
1        3.0  
2        1.0  
3        7.0  
4        3.0

### Preparing Year 2019 Data Set and Merging with Combined Data Set

#### Load the year 2019 data set and select relevant columns

Let's load the 2019's data set from `CPI2019-1.xlsx`.

In [ ]:
# Load year 2019 data set
CPI_2019 = pd.read_excel("raw_datasets/CPI2019-1.xlsx", sheet_name="CPI2019", header=2)

# Review the data set
print(CPI_2019.shape)
CPI_2019

As always, the number of rows matches the 180 country limit, and the headers and values look fine at least on a cursory look.

I'll also keep the following columns and rename them:

- Country
- ISO3
- Region
- CPI score 2019
- Rank

In [ ]:
# Pick the first 5 columns
CPI_2019 = CPI_2019.iloc[:, :5]

# Rename some of the columns
CPI_2019.rename(columns={ "Rank": "Rank 2019", "CPI score 2019": "CPI Score 2019"}, 
                inplace=True)

# Review modified data frame
CPI_2019

#### Merge the year 2019 data set with the combined data set

I'll merge the year 2019 data set following the same logic as the previous merge i.e. 2021 merged with 2020.

In [ ]:
# Merge year 2019 with combined data set
CPI = CPI.merge(CPI_2019, how="outer", on=["Country", "ISO3", "Region"])
CPI

### Preparing Year 2018 Data Set and Merging with Combined Data Set

#### Load the year 2018 data set and select relevant columns

Let's load the 2018 data set from `CPI2018_Full-Results_1801.xlsx`.

In [ ]:
# Load year 2018 data set
CPI_2018 = pd.read_excel("raw_datasets/CPI2018_Full-Results_1801.xlsx", sheet_name="CPI2018", header=2)

# Review the data set
print(CPI_2018.shape)
CPI_2018

As always, the number of rows matches the 180 country limit, and the headers and values look fine at least on a cursory look.

I'll also keep the following columns and rename them (if necessary):

- Country
- ISO3
- Region
- CPI Score 2018
- Rank

In [ ]:
# Pick the first 5 columns
CPI_2018 = CPI_2018.iloc[:, :5]

# Rename one of the columns
CPI_2018.rename(columns={ "Rank": "Rank 2018"}, 
                inplace=True)

# Review modified data frame
CPI_2018

#### Merge the year 2018 data set with the combined data set

I'll merge the year 2018 data set following the same logic as the previous merges.

In [ ]:
# Merge year 2018 with combined data set
CPI = CPI.merge(CPI_2018, how="outer", on=["Country", "ISO3", "Region"])
CPI_2018.rename(columns={ "Rank": "Rank 2018"}, 
                inplace=True)

The merge was mostly successful but oddly, the "Rank" column was not renamed to "Rank 2018". The first thing I can do is inspect the column names of the year 2018 data set.

In [ ]:
# Show all column names for the year 2018 data set
CPI_2018.columns

There's an extra white space after 'Rank'. The lesson here moving forward is I'll need to strip leading and trailing white spaces from the column names before renaming and merging.

In [ ]:
# Rename again but this time taking into account the trailing white space
CPI.rename(columns={ "Rank ": "Rank 2018"}, 
                inplace=True)
CPI.columns

### Preparing Year 2017 Data Set and Merging with Combined Data Set

#### Load the year 2017 data set and select relevant columns

Let's load the 2017 data set from `CPI2017_Full_DataSet-1801.xlsx`.

In [ ]:
# Load year 2017 data set
CPI_2017 = pd.read_excel("raw_datasets/CPI2017_Full_DataSet-1801.xlsx", sheet_name="CPI 2017", header=2)

CPI_2017.columns = CPI_2017.columns.str.strip()

# Review the data set
print(CPI_2017.shape)
CPI_2017

182 is more than the expected 180 country limit. It's possible that Transparency International extended the limit for this data set, but on further review I can see that there are two non-country rows. Both of them are the last two rows in the data set.

The first is just a row with all null/missing values; it's most likely an intentional empty row to separate the country rows from the global average row. The second is the global average row which calculates the average CPI score for all 180 countries in the year 2017; it is is not necessary in this project because I'll do a similar calculation in this project.

In [ ]:
# Keep all rows except the last two
CPI_2017 = CPI_2017.iloc[:-2]

print(CPI_2017.shape[0])
CPI_2017.tail()

Now that the two non-country rows have been removed, the number of rows now matches the expected 180 number of countries.

Next is selecting only the columns needed for the project:
- Country
- ISO3
- Region
- CPI Score 2017
- Rank

In [ ]:
# Pick the first 5 columns
CPI_2017 = CPI_2017.iloc[:, :5]
print(CPI_2017.columns)
# Rename one of the columns
CPI_2017.rename(columns={ "Rank": "Rank 2017"}, 
                inplace=True)

# Review modified data frame
CPI_2017

#### Merge the year 2017 data set with the combined data set

I'll merge the year 2017 data set following the same logic as the previous merges.

In [ ]:
# Merge year 2017 with combined data set
CPI = CPI.merge(CPI_2017, how="outer", on=["Country", "ISO3", "Region"])
CPI

### Preparing Year 2016 Data Set and Merging with Combined Data Set

#### Load the year 2016 data set and select relevant columns

Let's load the 2016 data set from `CPI2016_Results.xlsx`.

In [ ]:
import warnings

# Silence Conditional Formatting extension warning
warnings.simplefilter("ignore")

# Load year 2016 data set
CPI_2016 = pd.read_excel("raw_datasets/CPI2016_Results.xlsx", sheet_name="CPI2016_FINAL_16Jan")

# Re-enable the warning to default; it's possible for the other data sets to have an actual problem
warnings.simplefilter("default")

# 
CPI_2016.columns = CPI_2016.columns.str.strip()

# Review the data set
print(CPI_2016.shape)
CPI_2016

Year 2016 is the last year Transparency International did not standardize the 180 country limit. We will see similar patterns in the earlier years as well. The 176 number of rows is expected and it corresponds to the following excerpt from the [CPI 2016 page](https://www.transparency.org/en/cpi/2016):

> Over two-thirds of the **176 countries and territories** in this year's index fall below the midpoint of our scale of 0 (highly corrupt) to 100 (very clean).

Even though the data seems fine, there's actually a  Conditional Formatting extension warning that I've already silenced. [Conditional Formatting](https://support.microsoft.com/en-us/office/use-conditional-formatting-to-highlight-information-fed60dfa-1d3f-4e13-9ecb-f1951ff89d7f) is not relevant for this project because I'm not viewing the sheet directly nor am I going to highlight anything in the data frame according to the conditional formatting rules. Other than that, no Conditional Formatting rules in the sheet are used to provide information relevant for this project.

**From ISO3 to WB code**

Next, I'll need to point out another significant difference in the 2016 data set compared to the later years: countries are given a WB code instead of ISO3. The ISO3, or specifically [ISO 3166-1 alpha-3](https://www.iso.org/iso-3166-country-codes.html), is a standard used to represent a country with three-letter country codes. WB code, which presumably means World Bank code, is a term I wasn't able to learn more about even after reading the methodology documentations (refer: [Short](https://images.transparencycdn.org/images/CPI_2016_ShortMethodologyNote_EN.pdf) and [Technical](https://images.transparencycdn.org/images/CPI_2016_TechnicalMethodologyNote_EN.pdf)) for the data set.

The closest source would be this [Country Codes](https://wits.worldbank.org/wits/wits/witshelp/content/codes/country_codes.htm) which mentions the following:

> The country code table includes the WITS System country names for statistical purposes and both the **International Standards Organization (ISO) 3-digit alphabetic codes** and the United Nations Statistics Division (UNSD) 3-digit equivalent numeric codes. The names and codes are used in all of the three databases. 

The above implies WB code is exactly the same as ISO3. But I'm not confident that that's the case, so I would like to confirm that if there's no difference between the two at least in terms of Transparency International classification.

Before that, let's select only the following columns and rename them accordingly:

- Country
- Region
- CPI2016
- Rank
- WB code


In [ ]:
# Pick the first 5 columns
CPI_2016 = CPI_2016.iloc[:, :5]
print(CPI_2016.columns)
# Rename some of the columns
CPI_2016.rename(columns={ "Rank": "Rank 2016", "CPI2016": "CPI Score 2016"}, 
                inplace=True)
# Review modified data frame
CPI_2016

**Preliminary Merging to Investigate ISO3 and WB code difference**

Let us see if there's any difference between ISO3 and WB code.


In [ ]:
# Merge year 2016 with combined data set
ISO3_WB_comparison = CPI.merge(CPI_2016, how="outer", on=["Country", "Region"])
iso3_wb_mask = ISO3_WB_comparison['ISO3'] != ISO3_WB_comparison['WB Code']
ISO3_WB_comparison[iso3_wb_mask][['ISO3', 'WB Code']]

There is a lot to unpack here. I'm going to split the observations into multiple categories:

1. Both ISO3 and WB code are available
2. Only ISO3 is available
3. Only WB code is available

**Both ISO3 and WB code are available**

There are cases that even if both codes are available, the codes do not match each other. This proves that the ISO3 and WB Code do not necessarily agree with each other.

In [ ]:
# Isolate rows with which ISO3 and WB Code values are available
iso3_wb_both_available_mask = ISO3_WB_comparison['ISO3'].notnull() & ISO3_WB_comparison['WB Code'].notnull()
both_available = ISO3_WB_comparison[iso3_wb_both_available_mask][['ISO3', 'WB Code']]

# Get those rows with non-matching ISO3 and WB Code
iso3_wb_both_not_match_mask = both_available['ISO3'] != both_available['WB Code']
both_available[iso3_wb_both_not_match_mask]

For the above cases, I'll only use the ISO3 value and ignore the WB code in the combined data frame. Though, I will reserve that transformation later when I have surveyed all categories.

**Only ISO3 is available**

When this happens, it means that the countries are available in the 2021 - 2017 data sets but not available in the 2016 data set.

In [ ]:
# Isolate rows with which only ISO3 is available
iso3_wb_iso3_available_mask = ISO3_WB_comparison['ISO3'].notnull() & ISO3_WB_comparison['WB Code'].isnull()
ISO3_available = ISO3_WB_comparison[iso3_wb_iso3_available_mask][['ISO3', 'WB Code']]

# Review those rows with only ISO3 available
print(ISO3_available.shape[0])
ISO3_available[['ISO3', 'WB Code']]

There are 21 countries with only ISO3 values. It is okay to leave them as is because I do not need to reconcile any difference in country codes. Other than that, I am going to stick with ISO3 because that's the standard Transparency International adheres to in their recent data sets.

**Only WB Code is available**

The rows in this category requires further inspection. Because I am using ISO3, I will need to check if the WB Code corresponds to the current ISO3 code. If it does not, I'll have to replace the WB code with the appropriate ISO3 code.

In [ ]:
# Isolate rows with which only WB Code is available
iso3_wb_wb_code_available_mask = ISO3_WB_comparison['ISO3'].isnull() & ISO3_WB_comparison['WB Code'].notnull()
WB_available = ISO3_WB_comparison[iso3_wb_wb_code_available_mask][['ISO3', 'WB Code']]

# Review those rows with only WB Code available
print(WB_available.shape[0])
WB_available[['ISO3', 'WB Code']]

The result is a bit interesting. When I reviewed the "Only ISO3 is available", there is a row with the "USA" ISO3 code. Yet, at the same time, there is another row with the "USA" WB code. Why aren't they on the same row? I would like to investigate that issue further but I fear the cleaning needed might be too extensive for this section. I'll reserve extensive cleaning for now until everything has been merged.

For the time being, I'm going to let the data set as is. Other than that, moving forward, I'm not going merge data sets based on ISO3 matches and will stick to the Country and Region matches.

In [ ]:
# Merge the combined data set with the year 2016 data set based only on Country and Region
CPI = CPI.merge(CPI_2016, how="outer", on=["Country", "Region"])
CPI

### Preparing Year 2015 Data Set and Merging with Combined Data Set

**Load the year 2015 data set and select relevant columns**

Let's start by loading the 2016 data set from `CPI_2015_FullDataSet_2022-01-18-145020_enyn_2022-01-20-180010_mabu.xlsx`.

In [ ]:
# Load year 2015 data set
CPI_2015 = pd.read_excel("raw_datasets/CPI_2015_FullDataSet_2022-01-18-145020_enyn_2022-01-20-180010_mabu.xlsx", sheet_name="CPI 2015", header=0)

CPI_2015.columns = CPI_2015.columns.str.strip()

# Review the data set
print(CPI_2015.shape)
CPI_2015

There are only 168 countries in the year 2015 data set and that is expected as evident [here](https://www.transparency.org/en/press/corruption-perceptions-index-2015-corruption-still-rife-but-2015-saw-pocket):

> Overall, two-thirds of the **168 countries** on the 2015 index scored below 50, on a scale from 0 (perceived to be highly corrupt) to 100 (perceived to be very clean).

The first thing that caught my eye is that the values in Region are all missing (except for South Sudan's region). I used the Numbers spreadsheet editor to open the file and confirmed that the Regions are mostly empty. There's also an error for each cell which were expected to have the region value:

> The formula couldn't be imported because it contained a reference to another document. The formula was replaced by the last calculated value. Original formula: =VLOOKUP(C2, [1]!Table1[[wbcode]:[Region]], 2, FALSE)

The `VLOOKUP`([reference](https://support.microsoft.com/en-us/office/vlookup-function-0bbc8083-26fe-4963-8ab8-93a18ad188a1)) is an Excel function that allows a cell to refer to values in other cells either from the same or different document. It seems that there should be another spreadsheet document called Table1 but it was not included with the download. Then again, this is not a signficant problem and data for regions are easily accessible. In the meantime, I'll check first if South Sudan already exists in the combined data set. If it does, I can ignore the year 2015 region column and merge the data sets based only on Country name.

In [ ]:
# Check if South Sudan already exists in the combined data set
CPI[CPI["ISO3"] == "SSD"]

It exists so the Region column can be ignored. 

Continuing on, the year 2015 data set doesn't specify the country code standard it uses and unhelpfully named the column as "Country Code". But judging by the `VLOOKUP` syntax (`VLOOKUP(C2, [1]!Table1[[**wbcode**]:[Region]], 2, FALSE)` the standard used is likely to be WB Code. I like to be cautious here and keep the Country Code column for now.

The columns needed for the project:
- Country/Territory
- Country Code
- CPI 2015 Score
- Country Rank

In [ ]:
# Pick only the relevant columns
CPI_2015 = CPI_2015[["Country/Territory", "Country Code", "CPI 2015 Score", "Country Rank"]]
print(CPI_2015.columns)

# Rename some of the columns
CPI_2015.rename(columns={ "Country/Territory": "Country", "CPI 2015 Score": "CPI Score 2015", "Country Rank": "Rank 2015"}, inplace=True)

# Review modified data frame
CPI_2015

#### Merge the year 2015 data set with the combined data set

For this merge, I'll only merge based on country name since the region column has been removed.

In [ ]:
# Merge year 2015 with combined data set
CPI = CPI.merge(CPI_2015, how="outer", on=["Country"])
CPI

### Preparing Year 2014 Data Set and Merging with Combined Data Set

#### Load the year 2014 data set and select relevant columns

Let's load the 2014 data set from `CPI2014_Results.xlsx`.

In [ ]:
# Load year 2014 data set
CPI_2014 = pd.read_excel("raw_datasets/CPI2014_Results.xlsx", sheet_name="CPI 2014")

CPI_2014.columns = CPI_2014.columns.str.strip()

# Review the data set
print(CPI_2014.shape)
CPI_2014

Observations

175 countries

Region has a different format.

Still uses Country Code

In [ ]:
import pandas as pd
pd.options.display.max_columns = None

# import CPI data from 2012-2021
CPI_2021 = pd.read_csv("datasets/CPI/CPI2021-Table 1.csv", header=2)
CPI_2020 = pd.read_csv("datasets/CPI/CPI2020-Table 1.csv", header=2)
CPI_2019 = pd.read_csv("datasets/CPI/CPI2019-Table 1.csv", header=2)
CPI_2018 = pd.read_csv("datasets/CPI/CPI2018-Table 1.csv", header=2)
CPI_2017 = pd.read_csv("datasets/CPI/CPI2017-Table 1.csv", header=2)
CPI_2016 = pd.read_csv("datasets/CPI/CPI2016-Table 1.csv")
CPI_2015 = pd.read_csv("datasets/CPI/CPI2015-Table 1.csv")
CPI_2014 = pd.read_csv("datasets/CPI/CPI2014-Table 1.csv")
CPI_2013 = pd.read_csv("datasets/CPI/CPI2013-Table 1.csv", header=1)
CPI_2012 = pd.read_csv("datasets/CPI/CPI2012-Table 1.csv")

# retrieve the Country, ISO3, Region, CPI score, and rank columns from all datasets
# 2021
CPI_2021 = CPI_2021.iloc[:, :5]
CPI_2021.rename(columns={ "Country / Territory": "Country", "Rank": "Rank 2021"}, 
                inplace=True)

# 2020
CPI_2020 = CPI_2020.iloc[:, :5]
CPI_2020.rename(columns={ "Rank": "Rank 2020"}, inplace=True)
CPI = CPI_2021.merge(CPI_2020, how="outer", on=["Country", "ISO3", "Region"])

# 2019
CPI_2019 = CPI_2019.iloc[:, :5]
CPI_2019.rename(columns={ "Rank": "Rank 2019"}, inplace=True)
CPI = CPI.merge(CPI_2019, how="outer", on=["Country", "ISO3", "Region"])

# 2018
CPI_2018 = CPI_2018.iloc[:, :5]
CPI_2018.rename(columns={ "Rank ": "Rank 2018", "CPI Score 2018": "CPI score 2018"}, 
                inplace=True)
CPI = CPI.merge(CPI_2018, how="outer", on=["Country", "ISO3", "Region"])


# 2017
CPI_2017 = CPI_2017.iloc[:, :5]
CPI_2017.rename(columns={ "Rank ": "Rank 2017", "CPI Score 2017": "CPI score 2017"}, 
                inplace=True)
CPI = CPI.merge(CPI_2017, how="outer", on=["Country", "ISO3", "Region"])

# 2016 - from this point on, the datasets have slightly different columns and values than the above datasets
CPI_2016 = CPI_2016.iloc[:, :4]
CPI_2016.rename(columns={ "CPI2016": "CPI score 2016", "Rank": "Rank 2016"}, inplace=True)
CPI = CPI.merge(CPI_2016, how="outer", on=["Country"], suffixes=["", "_2016"])
CPI.drop("Region_2016", axis=1, inplace=True)

# 2015
CPI_2015 = CPI_2015.iloc[:, :5]
CPI_2015.drop("Region", axis=1, inplace=True)
CPI_2015.rename(columns={ "CPI 2015 Score": "CPI score 2015", 
                         "Country Rank": "Rank 2015", 
                         "Country/Territory": "Country", 
                         "Country Code": "ISO3"
                        }, 
                        inplace=True)
CPI = CPI.merge(CPI_2015, how="outer", on=["Country"], suffixes=["", "_2015"])

# 2014
CPI_2014 = CPI_2014.iloc[:, :5]
CPI_2014.drop("Region", axis=1, inplace=True)
CPI_2014.rename(columns={ "CPI 2014": "CPI score 2014", 
                         "Country Rank": "Rank 2014", 
                         "Country/Territory": "Country", 
                         "Country Code": "ISO3"
                        }, 
                        inplace=True)
CPI = CPI.merge(CPI_2014, how="outer", on=["Country"], suffixes=["", "_2014"])

# 2013
CPI_2013 = CPI_2013.iloc[:, :7]
CPI_2013.drop(labels=["Region", "IFS Code", "Country Rank.1"], axis=1, inplace=True)
CPI_2013.rename(columns={ "CPI 2013 Score": "CPI score 2013", 
                         "Country Rank": "Rank 2013", 
                         "Country / Territory": "Country", 
                         "WB Code": "ISO3"}, 
                         inplace=True)
CPI = CPI.merge(CPI_2013, how="outer", on=["Country"], suffixes=["", "_2013"])

# 2012
CPI_2012 = CPI_2012.iloc[:, :4]
CPI_2012.drop(labels=["Unnamed: 2"], axis=1, inplace=True)
CPI_2012.rename(columns={ "CPI 2012 Score": "CPI score 2012", 
                         "Country Rank": "Rank 2012", 
                         "Country / Territory": "Country"}, 
                         inplace=True)
CPI = CPI.merge(CPI_2012, how="outer", on=["Country"], suffixes=["", "_2012"])


# sanity check
print(CPI.shape)
CPI.head()

### Save Intermediary Data Sets

In [ ]:
# CPI_2013
CPI_2013t = pd.read_csv("datasets/CPI/CPI2013-Table 1.csv", header=1)
CPI_2013t

### Remove null values and duplicate rows

In [ ]:
# there are some null values in the final dataframe
# let's see if we can clean them up

# Country's frequency table
# No NaN
CPI["Country"].value_counts(dropna=False)

In [ ]:
# Country's columns
CPI.columns

In [ ]:
# We discovered some duplicate ISO3 columns
#
# The ISO3 code for certain countries has changed some time in between 2012 to 2021
# e.g. Montenegro's ISO3 was once MON but it changed to MNE after the CPI 2015 data was published
# Trivia: actually, the ISO3 change for Montenegro happened in 2007, but the CPI dataset still used the old ISO3 code
#
# We'll keep the newest ISO3 and drop the rest
CPI = CPI.drop(["ISO3_2015", "ISO3_2014", "ISO3_2013"], axis=1)
CPI.columns

In [ ]:
# There's a null value we haven't checked
# The curious thing is the value_counts doesn't show the NaN value
CPI["Country"].value_counts(dropna=False)

# we'll directly retrieve it with isnull
CPI[CPI["Country"].isnull()]

In [ ]:
# Row with index 186 has NaN for all its columns. 
# It probably came from an empty row in one of the original datasets. We can safely drop it.
CPI.drop(186, axis=0, inplace=True)

# 202 countries left
CPI.info()

In [ ]:
# We also discovered a country named GLOBAL AVARAGE below the previous row with index 186.
# There's barely any data in the GLOBAL AVARAGE and we don't need any of them. We decided to drop it.
CPI.drop(187, axis=0, inplace=True)

# Dropping the columns didn't reset index; let's do that
CPI.reset_index(drop=True, inplace=True)

In [ ]:
# There are some other peculiarities with the Country columns.
# The USA has multiple rows representing it because their country names differ:
# 1. United States of America
# 2. The United States of America
# 3. United States

CPI[CPI["Country"].str.match(r"(The )?United States")]

In [ ]:
# We'll stick with the row that has the most recent data ,i.e. row 27 with the name United States of America, 
# and fill the missing values using data from the older rows.
#
# This is a recurring issue, so it pays to create a function for it.

# Replace values of row with keep_indexes based on copy_dicts.
# Finish by dropping the copied rows and reset index.
def clean_inconsistent_names(df, keep_indexes, copy_dicts):
    # iterate over both keep_indexes and copy_dicts
    # the iterator variable for keep_indexes will be an index number
    # the iterator variable for copy_dicts will be a dictionary containing an index number mapped to the column names to copy from
    for keep_index, copy_dict in zip(keep_indexes, copy_dicts):
        for copy_index, cols in copy_dict.items():
            # this is where the copying happens 
            # i.e. a row with index of keep_index has its values replaced by values from a row with index of copy_index
            CPI.loc[keep_index, cols] = CPI.loc[copy_index, cols]
    
    drop_rows = []
    for dict in copy_dicts:
            # get the row index to drop
            drop_rows += dict.keys()
            
    drop_rows = list(drop_rows)
    CPI.drop(drop_rows, axis=0, inplace=True)
    CPI.reset_index(drop=True, inplace=True)

copy_usa_dicts = {186: ["CPI score 2016", "Rank 2016"], 
                  197: ["CPI score 2015", "Rank 2015", 
                        "CPI score 2014", "Rank 2014", 
                        "CPI score 2013", "Rank 2013", 
                        "CPI score 2012", "Rank 2012"]}
clean_inconsistent_names(CPI, [27], [copy_usa_dicts])

# we'll also drop the two rows and reset the index afterwards
CPI[CPI["Country"].str.match(r"(The )?United States")]

In [ ]:
# Country has no null and duplicate values as of now
CPI.info()
CPI[CPI.duplicated(subset="Country")]

In [ ]:
# ISO3 has a some null values as well
# We'll also be looking for duplicates at the same time
CPI[CPI["ISO3"].isnull()]

In [ ]:
# While reviewing the ISO3 duplicates, we discovered Country duplicates we didn't manage to capture before this
# e.g. Côte d’Ivoire is slightly different than Côte d´Ivoire because it has a Greek oxia character in place of apostrophe
#
# Other than that, there is another country with the same naming issue as the United States — Republic of Congo
#
# South and North Korea possibly have inconsistent names as well
#
# We'll start by examining both the Korea countries
CPI[CPI["Country"].str.contains("Korea")]

In [ ]:
# Copy data to most recent Korea countries rows
skorea_clean_dict = {189:  ["CPI score 2016", "Rank 2016", "CPI score 2015", "Rank 2015", 
             "CPI score 2014", "Rank 2014", 
             "CPI score 2013", "Rank 2013", 
             "CPI score 2012", "Rank 2012"]}

nkorea_clean_dict = {195:  ["CPI score 2016", "Rank 2016", "CPI score 2015", "Rank 2015", 
             "CPI score 2014", "Rank 2014", 
             "CPI score 2013", "Rank 2013", 
             "CPI score 2012", "Rank 2012"]}

# Clean both Korean countries
clean_inconsistent_names(CPI, [31, 174], [skorea_clean_dict, nkorea_clean_dict])

CPI[CPI["Country"].str.contains("Korea")]

In [ ]:
# we'll target Côte d’Ivoire next
CPI[CPI["Country"].str.contains("Ivoire")]

In [ ]:
# For unknown reasons, the clean_inconsistent_names function doesn't work well for Côte d’Ivoire
# So, we'll resort to copying the data without it

CPI.loc[181, ["CPI score 2017", "Rank 2017"]] = CPI.loc[185, ["CPI score 2017", "Rank 2017"]]

CPI.loc[181, ["CPI score 2016", "Rank 2016"]] = CPI.loc[190, ["CPI score 2016", "Rank 2016"]]

CPI.loc[181, ["CPI score 2015", "Rank 2015", 
              "CPI score 2014", "Rank 2014", 
              "CPI score 2013", "Rank 2013", 
              "CPI score 2012", "Rank 2012"]] = CPI.loc[194, ["CPI score 2015", "Rank 2015", 
                                                              "CPI score 2014", "Rank 2014", 
                                                              "CPI score 2013", "Rank 2013", 
                                                              "CPI score 2012", "Rank 2012"]]

CPI.drop([185,190,194], inplace=True)

CPI[CPI["Country"].str.contains("Côte")]

In [ ]:
# Next, Congo
CPI[CPI["Country"].str.match(r".*Congo")]

In [ ]:
# There are two countries that has "Congo" in their names
# On top of that, both of them also contain "Republic"
# The only thing that differentiates them is the word "Democratic"
#
# But, the recent CPI datasets (2021-2017) simply calls the Republic of the Congo as Congo, while the Democratic Republic of Congo retains the full name
# This is probably because they are using the country's ISO short name
#
# References: 
# Republic of the Congo - https://www.iso.org/obp/ui/#iso:code:3166:CG
# Democratic Republic of the Congo - https://www.iso.org/obp/ui/#iso:code:3166:CD
#
# To make the distinction clearer, we'll use their full name instead

# Rename most recent Republic of the Congo row to its full name
CPI.loc[161, ["Country"]] = "Republic of the Congo"
CPI[CPI["Country"].str.match(r".*Congo")]

In [ ]:
# Index 195 country name — Congo Republic — is ambiguous, 
# and because we don't have its ISO3 due to datasets merging, we should check its value in the original dataset
CPI_2015[CPI_2015["Country"].str.match(r".*Congo")]

In [ ]:
# We are now certain that index 195 refers to Republic of the Congo (the earlier datasets follow the same pattern)
CPI_2014[CPI_2014["Country"].str.match(r".*Congo")]

In [ ]:
# We'll clean Republic of the Congo first:
# index 161 - keep
# index 195 - copy 2013-2015 scores and ranks
# index 192 - copy 2016 score and rank

clean_rotc_dicts = {
                    195: ["CPI score 2015", "Rank 2015", 
                          "CPI score 2014", "Rank 2014", 
                          "CPI score 2013", "Rank 2013",
                          "CPI score 2012", "Rank 2012"],
                    192: ["CPI score 2016", "Rank 2016"]
                   }

clean_inconsistent_names(CPI, [161], [clean_rotc_dicts])
CPI[CPI["Country"].str.match(r".*Congo")]

In [ ]:
# Now, Democratic Republic of the Congo
# index 169 - keep
# index 189 - copy 2016 score and rank
clean_drotc_dict = {
                    189: ["CPI score 2016", "Rank 2016"]
                   }

clean_inconsistent_names(CPI, [169], [clean_drotc_dict])
CPI[CPI["Country"].str.match(r".*Congo")]

In [ ]:
# There doesn't seem to be any duplicate countries based on the following query
# Still, it's worthwhile reviewing each of the country to avoid missed duplicates
CPI[CPI["ISO3"].isnull()]

In [ ]:
# Checking the six ISO3 null countries if they are duplicates
# 1. Saint Vincent and The Grenadines
#
# Duplicate exist because of "the" and "The" mismatch
CPI[CPI["Country"].str.match(r".*Vincent")]

In [ ]:
# Clean Saint Vincent and the Grenadines
# We'll keep the row with the most recent data and copy the missing data from the row to be dropped
clean_inconsistent_names(CPI,[37],[{185: ["CPI score 2016", "Rank 2016"]}])
CPI[CPI["Country"].str.match(r".*Vincent")]

In [ ]:
# 2. Cape Verde
CPI[CPI["Country"].str.match(r".*Verde")]

In [ ]:
# Clean Cape Verde
# We'll keep the row with the most recent data and copy the missing data from the row to be dropped
clean_inconsistent_names(CPI,[38],[{185: ["CPI score 2016", "Rank 2016",
                                          "CPI score 2015", "Rank 2015",
                                          "CPI score 2014", "Rank 2014",
                                          "CPI score 2013", "Rank 2013",
                                          "CPI score 2012", "Rank 2012",]
                                   }
                                  ])
CPI[CPI["Country"].str.match(r".*Verde")]

In [ ]:
# 3. Brunei
# The name is longer for the row with the most recent data
CPI[CPI["Country"].str.match(r".*Brunei")]

In [ ]:
# Clean Brunei
# We'll keep the row with the most recent data and copy the missing data from the row to be dropped
clean_inconsistent_names(CPI,[180],[{185: ["CPI score 2016", "Rank 2016", 
                                          "CPI score 2013", "Rank 2013",
                                          "CPI score 2012", "Rank 2012",]
                                   }
                                  ])
CPI[CPI["Country"].str.match(r".*Brunei")]

In [ ]:
# 4. Macedonia
# Multiple names
CPI[CPI["Country"].str.match(r".*Macedonia")]

In [ ]:
# Clean Macedonia
# We'll keep the row with the most recent data and copy the missing data from the row to be dropped
clean_inconsistent_names(CPI,[91],[{184: ["CPI score 2018", "Rank 2018", 
                                          "CPI score 2017", "Rank 2017",
                                          "CPI score 2013", "Rank 2013",
                                          "CPI score 2012", "Rank 2012"
                                         ],
                                   
                                   185: [ "CPI score 2016", "Rank 2016",
                                           "CPI score 2015", "Rank 2015",
                                           "CPI score 2014", "Rank 2014"
                                         ]
                                       
                                   }
                                  ])
CPI[CPI["Country"].str.match(r".*Macedonia")]

In [ ]:
# 5 Guinea-Bissau
# No '-' for the most recent data
CPI[CPI["Country"].str.match(r".*(Bissau)")]

In [ ]:
# Clean Bissau
# We'll keep the row with the most recent data and copy the missing data from the row to be dropped
clean_inconsistent_names(CPI,[162],[{184: ["CPI score 2016", "Rank 2016", 
                                          "CPI score 2015", "Rank 2015",
                                          "CPI score 2014", "Rank 2014",
                                          "CPI score 2013", "Rank 2013",
                                          "CPI score 2012", "Rank 2012"
                                         ]
                                   }
                                  ])
CPI[CPI["Country"].str.match(r".*Bissau")]

In [ ]:
# 6 Puerto Rico
# Possibly the only country which missing ISO3 and has no duplicates
CPI[CPI["Country"].str.match(r".*(Puerto|Rico)")]

In [ ]:
# Puerto Rico's ISO3 is PRI — https://www.iso.org/obp/ui/#iso:code:3166:PR
CPI.loc[184, ["ISO3"]] = "PRI"
CPI[CPI["Country"].str.match(r".*(Puerto|Rico)")]

In [ ]:
# Confirm no nulls for ISO3
CPI[CPI["ISO3"].isnull()]

In [ ]:
# Confirm if all duplicates are dealt with
CPI[CPI["ISO3"].duplicated()]

In [ ]:
# We missed one duplicate because we didn't consider a case where Cote has no circumflex 'o'
CPI[CPI["ISO3"] == "CIV"]



In [ ]:
# We'll keep the row with the most recent data and copy the missing data from the row to be dropped
clean_inconsistent_names(CPI,[104],[{181: ["CPI score 2020", "Rank 2020",
                                           "CPI score 2017", "Rank 2017",
                                           "CPI score 2016", "Rank 2016",
                                          "CPI score 2015", "Rank 2015",
                                          "CPI score 2014", "Rank 2014",
                                          "CPI score 2013", "Rank 2013",
                                          "CPI score 2012", "Rank 2012"
                                         ]
                                   }
                                  ])

CPI[CPI["ISO3"] == "CIV"]

In [ ]:
CPI[CPI["ISO3"] == "CZE"]

In [ ]:
# We'll keep the row with the most recent data and copy the missing data from the row to be dropped
clean_inconsistent_names(CPI,[48],[{181: [
                                          "CPI score 2019", "Rank 2019",
                                          "CPI score 2018", "Rank 2018",
                                          "CPI score 2017", "Rank 2017",
                                          "CPI score 2016", "Rank 2016", 
                                          "CPI score 2015", "Rank 2015",
                                          "CPI score 2014", "Rank 2014",
                                          "CPI score 2013", "Rank 2013",
                                          "CPI score 2012", "Rank 2012"
                                         ]
                                   }
                                  ])

CPI[CPI["ISO3"] == "CZE"]

In [ ]:
CPI[CPI["ISO3"] == "SWZ"]

In [ ]:
# We'll keep the row with the most recent data and copy the missing data from the row to be dropped
clean_inconsistent_names(CPI,[121],[{181: [
             
                                          "CPI score 2018", "Rank 2018",
                                          "CPI score 2017", "Rank 2017",
                                     
                                          "CPI score 2014", "Rank 2014",
                                          "CPI score 2013", "Rank 2013",
                                          "CPI score 2012", "Rank 2012"
                                         ]
                                   }
                                  ])

CPI[CPI["ISO3"] == "SWZ"]

## Region



In [ ]:
# Get rows with missing Region
CPI[CPI["Region"].isnull()]

In [ ]:
# Looking at past datasets, Puerto Rico is in the AME (The Americas) region
CPI.loc[181, ["Region"]] = "AME"
CPI[CPI["ISO3"] == "PRI"]
# CPI["Country"].str.match(r".*(Puerto|Rico)")
# CPI[CPI["Country"].str.match(r".*(Puerto|Rico)]

In [ ]:
# Confirm no more null regions
CPI[CPI["Region"].isnull()]

### CPI scores and ranks

In [ ]:
CPI.info()

# Transparency International is responsible for collecting the data.
# They started standardizing the number of countries included in the global ranking to 180 countries each year in 2017.
#
# The numbers of non-null scores and ranks for the year 2016 and earlier matched the number of countries in the corresponding dataset
# e.g. 2016 dataset ranked 176 countries which corresponds to the 176 non-null CPI scores in our merged dataset
#
# If a country has a null score or rank for a particular year, it's either they don't pass the requirement to get included in the index,
# or their ranks are lower than the ranking threshold 
# i.e. only those ranked 180 or better are included in the later datasets, while the limit is different but still close to 180 in earlier dataset

In [ ]:
# As a precaution, we'll check whether the number of rows with non-null score and rank match the number of rows in the original dataset
# We only use the number of rows with non-null CPI score because if CPI score is null, rank will also be null

merged_rownums    = [year for year in range(2012, 2017)]
original_datasets = [CPI_2012, CPI_2013, CPI_2014, CPI_2015, CPI_2016]

print("Is the number of non-null scores and ranks in the merged dataset equal to the row numbers of the original datasets?")
for year, dataset in zip(merged_rownums, original_datasets):
    merged_length = CPI[~CPI["CPI score {}".format(year)].isnull()].shape[0]
    original_length = dataset.shape[0]
    
    print("{}: {}".format(year, merged_length == original_length))

# Analysis

## 

In [ ]:
import matplotlib.pyplot as plt
import geopandas as gpd

# import world boundaries data from World Bank: https://datacatalog.worldbank.org/search/dataset/0038272/World-Bank-Official-Boundaries
world = gpd.read_file("datasets/WB_countries_Admin0.geojson",driver="GeoJSON")

In [ ]:
# Taiwan is missing from the World Bank boundaries
# We imported Taiwan's boundary from a different site: https://earthworks.stanford.edu/catalog/stanford-dz142zj5454
taiwan = gpd.read_file("datasets/taiwan_boundaries.json",driver="GeoJSON")

In [ ]:
world[world["NAME_EN"] == "Netherlands"]

In [ ]:
world.head()

In [ ]:
# Merge the world and CPI datasets
# The merged dataset is mainly used for plotting a choropleth map
map_cpi = world.merge(CPI, how="outer", left_on="ISO_A3", right_on="ISO3")
map_cpi.head()

In [ ]:
# Merge Taiwan and CPI datasets
taiwan_map_cpi = taiwan.merge(CPI, how="inner", left_on="iso", right_on="ISO3")
taiwan_map_cpi
# taiwan

In [ ]:
# 4 countries were not merged correctly
map_cpi[map_cpi["geometry"].isnull()]

In [ ]:
# We can drop the countries that failed to merge and merge them again individually
map_cpi = map_cpi.drop(index=[251,252,253,254])

# map_cpi[map_cpi["NAME_EN"].str.match(r".*(Norway)|(Kosovo)|(France)")]
# Verify drop
map_cpi[map_cpi["geometry"].isnull()]

In [ ]:
# Taiwan is not included because its boundary is already covered by China
missing_countries = ["Norway", "Kosovo", "France"]

# Generate an array of column names in which each year has its corresponding CPI score and Rank
def gen_col_names():
    col_names = []
    for num in reversed(range(12,22)):
        col_names.append("CPI score 20{}".format(num))
        col_names.append("Rank 20{}".format(num))
    col_names.append("Region")
    return col_names
col_names = gen_col_names()

# Fill the missing values for countries in missing_countries
for country in missing_countries:
    for col_name in col_names:
        map_cpi.loc[map_cpi["NAME_EN"] == country, [col_name]] = CPI[CPI["Country"] == country].loc[:, col_name].iloc[0]
        
map_cpi[map_cpi["NAME_EN"].isin(missing_countries)].loc[:,"NAME_EN":"Rank 2012"]

In [ ]:
fig, map_axes = plt.subplots(5,2, figsize=(15,15))

from mpl_toolkits.axes_grid1 import make_axes_locatable

year = 2012

for ax_tuple in map_axes:
    for ax in ax_tuple: 
        if year % 2 != 0:
            divider = make_axes_locatable(ax)
            cax = divider.append_axes("right", size="3%", pad=0.1)
            ax.set_axis_off()
            ax.set_title(year,fontsize=18)
            map_cpi.plot(column='CPI score {}'.format(year), 
                         ax=ax,  
                         legend=True, 
                         legend_kwds={'label': "CPI Score"}, 
                         cax=cax,
                         cmap='RdYlGn',
                         vmin=0,
                         vmax=100
                        );

        else:
            ax.set_axis_off()
            ax.set_title(year,fontsize=18)
            map_cpi.plot(column='CPI score {}'.format(year), 
                         ax=ax,
                         cmap='RdYlGn',
                         vmin=0,
                         vmax=100
                        );
        # plot Taiwan after the main map, so it can overlap China
        taiwan_map_cpi.plot(
                            column='CPI score {}'.format(year), 
                            ax=ax, 
                            cmap='RdYlGn',
                            vmin=0,
                            vmax=100
                            )
        year+=1
        

## Worldwide CPI Score From 2012 to 2021

![CPI Score From 2012 to 2021](pictures/cpi_changes_smooth.gif "CPI Score From 2012 to 2021")

GIF created with [EZGif](https://ezgif.com/maker)

In [ ]:
cpi_cols = [col_name for col_name in col_names if 'CPI' in col_name]

score_means = CPI.describe()[cpi_cols].loc['mean',:]
score_means = score_means.rename(lambda x: int(x.replace("CPI score", "")))
score_means

In [ ]:
plt.style.available

In [ ]:
plt.style.use('seaborn-v0_8')
plt.plot(score_means)
plt.xlim(2012,2021)
plt.ylim(0,100)
plt.yticks(range(0,101,5))
# Transparency International's average worldwide CPI score goal
plt.axhline(50, color='green', alpha=0.25, linestyle='--')

plt.grid(False)

plt.show()

In [ ]:
plt.plot(score_means)
plt.xlim(2012,2021)

In [ ]:
# Calculate percentage of countries that's above 50
abv_50_pct = {}
abv_50_num = {}

year = 2021
for col in cpi_cols:
#     print(col)
    num_countries = CPI[col].notna().sum()
    num_countries_abv50 = (CPI[col] >= 50).sum()
    abv_50_pct[year] = num_countries_abv50 / num_countries * 100
    abv_50_num[year] = num_countries_abv50
    year -= 1
abv_50_pct
abv_50_num
#     print(CPI[col].notna().sum())
#     abv_50_pct[2012] = CPI[col].sum() / len()

In [ ]:
# Number of countries above 50
# ugly x and y
plt.plot(list(abv_50_num.keys()), list(abv_50_num.values()))
plt.yticks(range(53,60,1))
plt.xticks(range(2012,2022))
plt.xlim((2012,2021))

In [ ]:
import numpy as np
# Let's look at the percentages with CPI score above 50 across the years
plt.plot(list(abv_50_pct.keys()), list(abv_50_pct.values()))
plt.yticks(np.arange(30,33.5,0.5))
plt.xticks(range(2012,2022))
plt.xlim((2012,2021))

In [ ]:
# create below 50
# create a function for this
# Calculate percentage of countries that's below 50
bel_50_pct = {}
bel_50_num = {}

year = 2021
for col in cpi_cols:
#     print(col)
    num_countries = CPI[col].notna().sum()
    num_countries_bel50 = (CPI[col] < 50).sum()
    bel_50_pct[year] = num_countries_bel50 / num_countries * 100
    bel_50_num[year] = num_countries_bel50
    year -= 1
bel_50_pct
bel_50_num

In [ ]:
# Number of countries above 50
# ugly x and y
plt.plot(list(bel_50_num.keys()), list(bel_50_num.values()))
# plt.yticks(range(114,124,1))
plt.xticks(range(2012,2022))
plt.xlim((2012,2021))

In [ ]:
import numpy as np
# Let's look at the percentages with CPI score above 50 across the years
plt.plot(list(bel_50_pct.keys()), list(bel_50_pct.values()))
# plt.yticks(np.arange(30,33.5,0.5))
plt.xticks(range(2012,2022))
plt.xlim((2012,2021))

In [ ]:
fig, ax = plt.subplots(figsize=(15,8))

x = np.arange(10)  # the label locations
width = 0.4  # the width of the bars

rev_abv_50_pct = list(reversed(list(abv_50_pct.values())))
rev_bel_50_pct = list(reversed(list(bel_50_pct.values())))

rects1 = ax.bar(x - width/2,  rev_abv_50_pct, width,  label='Men')
rects2 = ax.bar(x + width/2,  rev_bel_50_pct, width,  label='Women')

ax.bar_label(rects1, padding=4)
ax.bar_label(rects2, padding=4)
ax.set_xticks(x, list(range(2012,2022)))

plt.show()

### Region CPI score

In [ ]:
cpi_region_means = CPI.groupby("Region").mean(numeric_only=True)[cpi_cols]

In [ ]:
# AME   - The Americas
# AP    - Asia Pacific
# ECA   - Eastern Europe & Central Asia
# MENA  - Middle East & North Africa
# SSA   - Sub-Saharan Africa
# WE/EU - Western Europe

year_list = list(range(2012,2022))

for region in cpi_region_means.iterrows():
    reversed_cpi = list(reversed(region[1]))
    plt.plot(year_list, reversed_cpi, label=region[0])
    
plt.xlim(2012, 2021)
plt.ylim(0, 100)
plt.yticks(range(0,101,5))
plt.legend(loc='upper right')

In [ ]:
# Compare with main CPI dataframe
# print(map_cpi.groupby("Region").mean(numeric_only=True))
# print(CPI.groupby("Region").mean(numeric_only=True))
print(map_cpi.groupby("Region").size())
print(CPI.groupby("Region").size())

# It seems that the Netherlands has multiple polygons and all of them are not duplicates
# map_cpi[map_cpi["Region"] == "WE/EU"].duplicated(subset="geometry")
# map_cpi[map_cpi["Region"] == "WE/EU"].duplicated()

map_cpi[map_cpi["ISO3"] == "CIV"]
CPI[CPI["ISO3"] == "CIV"]

In [ ]:
map_cpi[map_cpi.duplicated(subset="ISO3") & ~map_cpi["ISO3"].isna()]

In [ ]:
# map_cpi[map_cpi["N"] == "WE/EU"]
map_cpi["FORMAL_EN"].value_counts()

In [ ]:
# The size of WE/EU in map_cpi is bigger than the size of WE/EU in map_cpi because the Netherlands has 3 extra geometries
map_cpi[map_cpi["NAME_EN"] == "Netherlands"]

In [ ]:
# New Zealand has two geometries yet the size of AP is still the same; it's because Taiwan is not in map_cpi thus balancing out the extra country
map_cpi[map_cpi["NAME_EN"] == "New Zealand"]

In [ ]:
CPI[CPI["Region"] == "AP"]["Country"].value_counts().sort_index()

In [ ]:
map_cpi[map_cpi["Region"] == "AP"]["Country"].value_counts().sort_index()
# map_cpi[map_cpi["Region"] == "AP"]

In [ ]:
CPI[CPI["Region"] == "WE/EU"]["Country"].value_counts().sort_index()

In [ ]:
map_cpi[map_cpi["Region"] == "WE/EU"]["Country"].value_counts().sort_index()

In [ ]:
region_cpi = CPI.groupby("Region").mean(numeric_only=True)[cpi_cols]
region_cpi = region_cpi.rename(lambda x: int(x.replace("CPI score", "")), axis=1)

In [ ]:
# plt.plot(region_cpi)
# plt.plot(region_cpi["AME"])
# plt.plot(region_cpi.loc["AME"].sort_index())
for region in region_cpi.index:
    print(region)
    plt.plot(region_cpi.loc[region], label=region)
    #     print(region)
#     print(data)
#     plt.plot(region)
# plt.plot(.loc["AME"])
# year
plt.legend()

region_cpi.index

In [ ]:
region_cpi

In [ ]:
# map_cpi["Region"].dropna().unique()
# map_cpi[map_cpi["Region"] == "AP"].plot()


map_cpi_filled_region = map_cpi.dropna(subset="Region")

def get_region_mean(row, region_means):
    row_region = row["Region"]
    for year in range(2012,2022):
        region_mean = region_means.loc[row_region]["CPI score {}".format(year)]
        col_name = "region_mean_{}".format(year)
        row[col_name] = region_mean
    return row

test=map_cpi_filled_region.copy()

test_2 = map_cpi_filled_region.apply(lambda row: get_region_mean(row, cpi_region_means), axis=1)
# test_2.iloc[:,-10:-1]

# ap_region = test_2[test_2["Region"] == "AP"]

# fig, ax = plt.subplots(1,1)


# print(norm(ap_region.iloc[0]["region_mean_2012"]))
# print(color)
# # ax.set_axis_off()
# ap_region.plot( 
# #                 column='region_mean_2012', 
#                 ax=ax,
#                 legend=True, 
#                 legend_kwds={'label': "CPI Score"},
#                 color = color,
# #                 cmap='RdYlGn',
# #                 vmin=0,
# #                 vmax=100
# )

# ax.set_title(year,fontsize=18)
# plt.xlim(50, 200)
# plt.show()

# test_2.plot( column='region_mean_2012', 
#                 ax=ax,
#                 legend=True, 
#                 legend_kwds={'label': "CPI Score"},
# #                 color = color,
#                 cmap='RdYlGn',
#                 vmin=0,
#                 vmax=100
#            )
# for region, cpi_scores in cpi_region_means.iterrows():
#     fig, map_axes = plt.subplots(1,10)
#     for ax in map_axes:
        
# cpi_region_means
# 10 default
# fig, map_axes = plt.subplots(2,1, figsize=(30,50))


# from mpl_toolkits.axes_grid1 import make_axes_locatable

year = 2012

years=list(range(2012,2022))

import math
from matplotlib import cm, colors
norm = colors.Normalize(vmin=0, vmax=100)

cmap = cm.get_cmap('RdYlGn')


for year in years:
    fig, ax = plt.subplots(1,1, figsize=(15,15))
    ax.set_axis_off()
    # TODO: Make a function for the following later
    # circle for WE/EU
    we_mean = math.floor(region_cpi.loc["WE/EU"][year])
    color = cmap(norm(we_mean))
    we_circle = plt.Circle((-15,76), 8, color="#3695d8")
    we_circle_border = plt.Circle((-15,76), 8, color="gray", alpha=0.5, fill=False, linewidth=1)
    # x axis + 11, y axis - 2 from main circle coordinate
    ax.text(-4, 74, "WE/EU", color='black', fontsize=11.5, fontweight='bold')
    # x axis - 4.2, y axis - 3 from main circle coordinate
    ax.text(-19.5,73, we_mean, fontsize=18, color='white', fontweight='bold')
    ax.add_patch(we_circle)
    ax.add_patch(we_circle_border)
    
   
    
    divider = make_axes_locatable(ax)
    cax = divider.append_axes("right", size="3%", pad=0.1)
#     ax.add_patch(circle1)
#     ax.add_patch(circle_border)
#         ax.set_axis_off()
    ax.set_title(year,fontsize=18)
    test_2.plot(column='region_mean_{}'.format(year), 
                     ax=ax,  
                     legend=True, 
                     legend_kwds={'label': "CPI Score"}, 
                     cax=cax,
                     cmap='RdYlGn',
                     vmin=0,
                     vmax=100
                    );

        
     # circle for AP
    ap_mean = math.floor(region_cpi.loc["AP"][year])
    ap_circle = plt.Circle((150,0), 8, color="#3695d8")
    ap_circle_border = plt.Circle((150,0), 8, color="gray", alpha=0.5, fill=False, linewidth=1)
    # x axis + 13, y axis - 2
    ax.text(161, -2, "AP", color='black', fontsize=11.5, fontweight='bold')
    # x axis - 4.2, y axis - 3
    ax.text(145.5,-3, ap_mean, fontsize=18, color='white', fontweight='bold')
    ax.add_patch(ap_circle)
    ax.add_patch(ap_circle_border)
    
    # circle for ECA aka Eastern Europe and Central Asia
    eca_mean = math.floor(region_cpi.loc["ECA"][year])
    circle_x = 150
    circle_y = 76
    circle_coord = (circle_x, circle_y)
    eca_circle = plt.Circle(circle_coord, 8, color="#3695d8")
    eca_circle_border = plt.Circle(circle_coord, 8, color="gray", alpha=0.5, fill=False, linewidth=1)
    # x axis + 13, y axis - 2
    ax.text(circle_coord[0]+13, circle_coord[1]-2, "ECA", color='black', fontsize=11.5, fontweight='bold')
    # x axis - 4.2, y axis - 3
    ax.text(circle_coord[0]-5,circle_coord[1]-3, eca_mean, fontsize=18, color='white', fontweight='bold')
    
    ax.add_patch(eca_circle)
    ax.add_patch(eca_circle_border)
    
     # circle for MENA aka Middle East and North Africa
    mena_mean = math.floor(region_cpi.loc["MENA"][year])
    circle_x = 40
    circle_y = 30
    circle_coord = (circle_x, circle_y)
    mena_circle = plt.Circle(circle_coord, 8, color="#3695d8")
    mena_circle_border = plt.Circle(circle_coord, 8, color="gray", alpha=0.5, fill=False, linewidth=1)
    # x axis + 13, y axis - 2
    ax.text(circle_coord[0]+11, circle_coord[1]-2, "MENA", color='black', fontsize=11.5, fontweight='bold')
    # x axis - 4.2, y axis - 3
    ax.text(circle_coord[0]-5,circle_coord[1]-3, mena_mean, fontsize=18, color='white', fontweight='bold')
    
    ax.add_patch(mena_circle)
    ax.add_patch(mena_circle_border)
    
     # circle for SSA or Sub-Saharan Africa
    ssa_mean = math.floor(region_cpi.loc["SSA"][year])
    circle_x = 25
    circle_y = 0
    circle_coord = (circle_x, circle_y)
    ssa_circle = plt.Circle(circle_coord, 8, color="#3695d8")
    ssa_circle_border = plt.Circle(circle_coord, 8, color="gray", alpha=0.5, fill=False, linewidth=1)
    # x axis + 13, y axis - 2
    ax.text(circle_coord[0]+11, circle_coord[1]-2, "SSA", color='black', fontsize=11.5, fontweight='bold')
    # x axis - 4.2, y axis - 3
    ax.text(circle_coord[0]-5,circle_coord[1]-3, ssa_mean, fontsize=18, color='white', fontweight='bold')
    
    ax.add_patch(ssa_circle)
    ax.add_patch(ssa_circle_border)
    
     # circle for AME or The Americas
    ame_mean = math.floor(region_cpi.loc["AME"][year])
    circle_x = -125
    circle_y = 0
    circle_coord = (circle_x, circle_y)
    ame_circle = plt.Circle(circle_coord, 8, color="#3695d8")
    ame_circle_border = plt.Circle(circle_coord, 8, color="gray", alpha=0.5, fill=False, linewidth=1)
    # x axis + 13, y axis - 2
    ax.text(circle_coord[0]+11, circle_coord[1]-2, "AME", color='black', fontsize=11.5, fontweight='bold')
    # x axis - 4.2, y axis - 3
    ax.text(circle_coord[0]-5,circle_coord[1]-3, ame_mean, fontsize=18, color='white', fontweight='bold')
    
    ax.add_patch(ame_circle)
    ax.add_patch(ame_circle_border)
    
    
    plt.savefig(f"pictures/regional_{year}", bbox_inches='tight')
    plt.show()
    year+=1
        
# test_2


![regional cpi changes](pictures/regional_cpi_changes_crossfade.gif "Regional CPI Changes 2012-2021")

### Compare 2012 to 2021 CPI score - which country changed and which didn't

In [ ]:
# For those countries with CPI score in 2012 and 2021, how many of those countries have a different CPI score in 2021 than in 2012
CPI["changes_CPI_12to21"] = CPI["CPI score 2021"] - CPI["CPI score 2012"]

In [ ]:
CPI[CPI["changes_CPI_12to21"] >= 1][["Country", "CPI score 2021", "CPI score 2012"]]

In [ ]:
# plot countries with at least one point improvement
cpi_change_threshold = 5
cpi_improvements = CPI[CPI["changes_CPI_12to21"] >= cpi_change_threshold][["Country", "CPI score 2021", "CPI score 2012", "changes_CPI_12to21"]]
cpi_declined     = CPI[CPI["changes_CPI_12to21"] <= -cpi_change_threshold][["Country", "CPI score 2021", "CPI score 2012", "changes_CPI_12to21"]]
cpi_stagnated         = CPI[CPI["changes_CPI_12to21"].between(-cpi_change_threshold,cpi_change_threshold)][["Country", "CPI score 2021", "CPI score 2012", "changes_CPI_12to21"]]
cpi_improvements

In [ ]:
cpi_declined

In [ ]:
cpi_stagnated

In [ ]:
# plot improved, declined, and 0
sum_improved  = cpi_improvements.shape[0]
sum_declined  = cpi_declined.shape[0]
sum_stagnated = cpi_stagnated.shape[0]


plt.bar(["Improved", "Declined", "Stagnated"],[sum_improved, sum_declined, sum_stagnated])

In [ ]:
# top 10 countries improved
cpi_improvements.sort_values("changes_CPI_12to21", ascending=False)[:10]

In [ ]:
# top 10 countries declined
cpi_declined.sort_values("changes_CPI_12to21")[:10]

### Credits

[Corruption Perceptions Index](https://www.transparency.org/en/cpi/) by Transparency International is licensed under CC-BYND 4.0.

### Region

The CPI column especially is important to rename because in the final data frame, the CPI for all years will be available i.e. each CPI column is identified by a year.

In [ ]:
# Code Graveyard

# import openpyxl as op

# wb = op.load_workbook("raw_datasets/CPI2016_Results.xlsx", data_only=True)

# from itertools import islice

# data = wb["CPI2016_FINAL_16Jan"].values
# cols = next(data)[1:]
# data = list(data)
# idx = [r[0] for r in data]
# data = (islice(r, 1, None) for r in data)
# CPI_2016 = pd.DataFrame(data, index=idx, columns=cols)
# CPI_2016